compiler/rustc_mir/src/interpret/intern.rs

```diff
@@ -104,7 +104,7 @@ fn intern_shallow<'rt, 'mir, 'tcx, M: CompileTimeMachine<'mir, 'tcx>>(
     // This match is just a canary for future changes to `MemoryKind`, which most likely need
     // changes in this function.
     match kind {
-        MemoryKind::Stack | MemoryKind::Vtable | MemoryKind::CallerLocation => {}
+        MemoryKind::Stack | MemoryKind::Heap | MemoryKind::Vtable | MemoryKind::CallerLocation => {}
```

Comment:
Should Miri also use this `Heap` variant? Or should we rename it to `ConstHeap`?

Reply:
ah, good point. I think miri will need its own heap variant to prevent us from accidentally mixing the two somewhere.

Reply:
@oli-obk so what do you think about using a non-`!` `MemoryKind` for the CTFE machine?

Reply:
Ah, apologis, forgot to post. @vn-ki experimented with that, but it's a nontrivial change, because `ConstProp` reuses a lot of the same machinery. I am still in favor of doing that change, but in a follow up PR that does *just* that single change

Reply:
if you are worried about miri breakage, we can also do that change to an `enum MemoryKind {}` in a PR and then rebase this PR on top of it.

Reply:
No I think this is fine (but please leave a FIXME in the `enum` definition).